### ISSUE TITLE
Fails cross-compilation on musl

### ISSUE THREAD

**Issue by vi** [indianakernick/renamore#1]
`/home/vi/.cargo/registry/src/-3d9d141e372ea94e/renamore-0.3.0/src/linux.rs:26: undefined reference to renameat'`

Shall it expose a cargo feature to force fallback and/or force usage of `syscall` or something like that without relying on C library?

**Comment by indianakernick** [comment]
From what I can tell, the musl standard library has `renameat`. It's [right here](https://git.musl-libc.org/cgit/musl/tree/src/unistd/renameat.c). Still, the platform detection could be improved. The `build.rs` could check for the existence of the symbols. Doing a direct `syscall` is probably not a good idea because the syscall number would depend on the architecture ([see the definitions in musl](https://git.musl-libc.org/cgit/musl/commit/?id=002e8d84c5fec47636606aef5d2f16914c6097f3)).

**Comment by vi** [comment]
My current workaround is to build this file:

```c
#define _GNU_SOURCE 1
#include <sys/syscall.h>
#include <unistd.h>

int renameat2(int oldfd, const char *old, int newfd, const char *new, unsigned int flags)
{
	return syscall(SYS_renameat2, oldfd, old, newfd, new, flags);
}
```

(based on `renameat` definition in Musl)

and use linker wrapper script in `~/.cargo/config`'s `[target.x86_64-unknown-linux-musl].linker`:

```sh
#!/bin/sh
cc "$@" /path/to/renameat2_amd64.o
```

---

So maybe if `build.rs`-based symbol detection for Musl fails to find `renameat2` then this workaround should be activated from within `renamore`?

**Comment by indianakernick** [comment]
Oh it's `renameat2` not `renameat`! If the syscall exists and musl isn't providing a wrapper then that sounds like a problem with musl. In this case, checking for the existence of the `renameat2` wrapper inside of `build.rs` wouldn't help you because you'd end up with the fallback even though the real thing exists.

So yeah (you edited your comment while I was writing mine!), this specific workaround for musl could be added so that the `build.rs` includes the above definition if the `target_env="musl"` and the symbol couldn't be found.

**Comment by vi** [comment]
> If the syscall exists and musl isn't providing a wrapper then that sounds like a problem with musl.

Musl seems to _use_ `renameat2` if it is available, but only _expose_ `renameat` function.

**Comment by indianakernick** [comment]
I pushed something up. I haven't tested it. Let me know if it works for you.

**Comment by vi** [comment]
Shall a special Cargo feature to turn off all `renamore` features (i.e. always fail `rename_exclusive`, always false `rename_exclusive_is_atomic`, always falls back in `rename_exclusive_fallback`) be exposed anyway, even with `musl` support being corrected?

This would allow users to continue building apps (in reduced mode) by specifying something like `--features renamore/force-fallback`.

**Comment by indianakernick** [comment]
Yeah I think so. It would be an easy workaround for future issues with platform detection.

**Comment by vi** [comment]
> I pushed something up. I haven't tested it. Let me know if it works for you.

It stops issuing exclusive renames on musl at all, even if I have my workaround active.
No mentions of `renameat2` symbol in `librenamore` rlib.

**Comment by indianakernick** [comment]
I found some time to do some further investigation into this and I think I know what's going on.

The host target is `aarch64-unknown-linux-gnu` and I'm compiling for `aarch64-unknown-linux-musl`. Constructing the command to invoke the compiler on the `test.c` file is failing because the compiler is not installed. This causes an early-exit and the `linker` cfg variable is not set. I installed the compiler with `sudo apt install musl-dev`. After that, the command was able to run and the compiler returned a non-zero exit code as expected. What I didn't expect was that `cfg!(target_env = "musl")` was `false` even when passing `--target aarch64-unknown-linux-musl`. So the `linux-musl.c` file isn't compiled. When I comment out the `if` and let it run unconditionally, the tests pass.

Thinking about it a little more, it actually makes total sense. The build script is built for the host. It needs to run on the host. So of course the cfg variables will describe the host. To get the true target, you need to use the `TARGET` environment variable. With that, it now seems to be working. Although I seem to be having some problems on Windows but that's a separate issue (oh Windows!).

Let me know if this is working in your situation.

**Comment by vi** [comment]
Trying with adbfa625bfb3242bef0739288b55d3adc1dff2f2:

* x86_64-unknown-linux-gnu: works
* x86_64-unknown-linux-musl + workaround: `multiple definition of renameat2; /mnt/fluff/rust-targets/duplo/x86_64-unknown-linux-musl/debug/deps/librenamore-7a95ced6ab76a441.rlib(linux-musl.o):/mnt/bkel/vi/src/renamore/src/linux-musl.c:5: first defined here`
* x86_64-unknown-linux-musl without workaround: works, `librenamore.rlib` defines `renameat2`.
* x86_64-unknown-linux-musl + workaround + `CC` environment variable points to the same compiler wrapper used for the workaround: works, `librenamore.rlib` does not define `renameat2`.

If non-Linux works properly then this version looks suitable for the `v0.3.1` release.

---

Maybe marking `renameat2` as a weak symbol can be helpful even with the auto-detection:

```diff
diff --git a/src/linux-musl.c b/src/linux-musl.c
index 8eb1234..712d131 100644
--- a/src/linux-musl.c
+++ b/src/linux-musl.c
@@ -2,6 +2,7 @@
 #include <sys/syscall.h>
 #include <unistd.h>

+__attribute__((weak, visibility("default")))
 int renameat2(int oldfd, const char *old, int newfd, const char *new, unsigned int flags) {
     return syscall(SYS_renameat2, oldfd, old, newfd, new, flags);
 }
```

This way it works even if `renameat2` is exported from `renamore` by mistake.

**Comment by indianakernick** [comment]
> If non-Linux works properly then this version looks suitable for the `v0.3.1` release.

Well that's the thing. It's fine on Linux and macOS but I haven't been able to get it to work on Windows. Although I've realised that the way I've approached this is a little silly.

 - Linux added `renameat2` in kernel version 3.15. The oldest supported LTS version of the Linux kernel is 4.14.
 - macOS added `renamex_np` in 10.12 (Sierra). The oldest supported macOS version is 11 (Big Sur).
 - Windows added `MoveFileExW` in Windows XP. The oldest supported LTS version of Windows is Windows 10.

I don't think anyone is creating new software for operating systems that aren't receiving security patches anymore. However, there is some special handling required on Linux due to the diversity of C standard library implementations. So I'll remove the macOS and Windows checks but leave the Linux code path as it is. Making `renameat2` a weak symbol makes sense for the sake of robustness so I'll do that.

With that, I'll probably release this later today (likely within 12 hours).
